tests/abi/abi.c

```diff
@@ -0,0 +1,175 @@
+#define DUMP_REGISTER(REGISTER)                           \
+  do {                                                    \
+    printf("    - Name: \"%s\"\n      Value: 0x%016lx\n", \
+           STRINGIFY(REGISTER),                           \
+           Generated[RegisterIndex--]);                   \
+  } while (0);
+  {
+    uint32_t RegisterIndex = REGISTER_COUNT - 1;
+    FOR_EACH(DUMP_REGISTER, REGISTER_LIST);
+  }
+#undef DUMP_REGISTER
+
+  /* Dump Stack */
+  puts("  Stack:");
+  for (uint32_t Index = REGISTER_COUNT; Index < GENERATED_COUNT; ++Index)
+    printf("    - Offset: -0x%04x\n      Value: 0x%016lx\n",
+           (GENERATED_COUNT - Index) * REGISTER_SIZE,
+           Generated[Index]);
+}
+
+#define CALL_ITERATION_WITH(FUNCTION_NAME) \
+  CONCATENATE(iterationImplOn_, FUNCTION_NAME)
+
+#define GENERATED_BYTES ((GENERATED_COUNT) * sizeof(REGISTER_C_TYPE))
+
+/* clang-format off */
+#define DEFINE_ITERATION_WITH(FUNCTION_NAME)                                 \
+void CALL_ITERATION_WITH(FUNCTION_NAME)(void) {                              \
+  REGISTER_C_TYPE *Generated = (REGISTER_C_TYPE *) malloc(GENERATED_BYTES);  \
+  generate(Generated);                                                       \
+  dump(Generated);                                                           \
+                                                                             \
+  __asm__ __volatile__ (                                                     \
+    /* Save all the registers */                                             \
+    FOR_EACH(PUSH_REGISTER, REGISTER_LIST)                                   \
+    /* Popularize the stack with random data */                              \
+    REPEAT(PUSH_CONSTANT, STACK_VALUE_COUNT, REGISTER_COUNT, FIRST_REGISTER) \
+    /* Popularize all the registers with random data */                      \
+    INDEXED_FOR_EACH(FILL_REGISTER, REGISTER_LIST)                           \
+    /* Call the function */                                                  \
+    CALL_A_FUNCTION(FUNCTION_NAME)                                           \
+    /* Restore the stack */                                                  \
+    RESTORE_STACK(STACK_VALUE_COUNT)                                         \
+    /* Restore all the registers */                                          \
+    REVERSE_FOR_EACH(POP_REGISTER, REGISTER_LIST)                            \
+    :: "X"(Generated)                                                        \
+    : COMMA_SEPARATED_FOR_EACH(STRINGIFY, REGISTER_LIST), "memory"           \
+  );                                                                         \
+                                                                             \
+  free(Generated);                                                           \
+}
+/* clang-format on */
+
+#ifdef __APPLE__
+#define FUNCTION_NAME_PREFIX _
+#else
+#define FUNCTION_NAME_PREFIX
+#endif
+
+#ifdef REG_
+#undef REG_
+#endif
```

Comment:
Why?

Reply:
It's a really dirty workaround I had to do because gcc's inline assembly sucks! When the capture-like thing is used, the registers have to use `%%` prefix, when it is not - `%`! And I have to have the capture thing for the assembly inside a c function. And I'm not allowed to have one in the "free" assembly block later down. So it's a way insert the whatever prefix is needed at a time in the right place.

Reply:
Can you see if you can solve the problem with an extra macro argument for macros using `REG_`?
Or maybe have two distinct macros.

Reply:
Problem is that everything is loops on top of loops. And this was a super-late addition to the fray, so I decided against rewriting everything from scratch just to accommodate for this (this was already taking way longer than it should have at that point, even more so now).

Reply:
AFAIU we could use the capture-like `__asm__` except for `getTheValue` (and `restoreState`).
Let's see if we can get rid/inline them and switch to only use `__asm__` with the capture-like thing.

Reply:
that would be the dream come true!

Reply:
Done, done, and done! And I couldn't be happier about that!